### PR TITLE
Attestation performance metrics

### LINKED ISSUE #7694: Add Inclusion Distance as prometheus metrics (maybe Attestation Effectiveness)
# 🚀 Feature Request

### Description

Being able to monitor our inclusion distance over time, and ideally calculate the attestation effectiveness of our validator node would help stakers noticing when there's potentially an issue with our node causing our attestations to be broadcast late OR worse, not at all.
An example for me, during this period of non-finality was that the default max-peers (3) was causing some attestations to never be included, and an avg. distance of 2.5. Increasing this helped substantially. 

### Describe the solution you'd like
add a `inclusion_distance` and an `attestation_effectiveness` metrics

### PR DESCRIPTION
**Pull request by potuz**
**What type of PR is this?**

Feature

Fixes #7694 

**What does this PR do? Why is it needed?**

This PR  exports 5 metrics by pubkey: the attested slot, the inclusion distance, the correctness of the vote on head,target and source. In order to be correct however this is now based on #7708 since otherwise it would suffer from #7699

**Other notes for review**

The attested slot is always one epoch ahead with this PR. This has to be taken into account when producing tables like this one

import

I have checked the above for over a day with beaconcha.in now that it is working.

### CODE REVIEW

File: validator/client/metrics.go [hunk 1]
```diff
@@ -85,6 +85,61 @@ var (
...
+		prometheus.GaugeOpts{
+			Namespace: "validator",
+			Name:      "correctly_voted_target",
+			Help:      "True if correctly voted target in last attestation.",
+		},
+		[]string{
+			"pubkey",
+		},
+	)
+	// ValidatorCorrectlyVotedHeadGaugeVec used to keep track of validator's accuracy on voting target by public key.
```
Comment (terencechain) [inline validator/client/metrics.go:132]:
```suggestion
	// ValidatorCorrectlyVotedHeadGaugeVec used to keep track of validator's accuracy on voting head by public key.
```

**Review by terencechain — CHANGES_REQUESTED** [review]
I'm slightly concerned on how this will increase the cardinality of the metrics page significantly and require tsdb to store much more data than before.

Is a gauge the most appropriate format?
We track overall inclusion distance with a histogram in the beacon node, should we use a histogram?

File: validator/client/metrics.go (hunk 1 above)
Comment (potuz) [reply validator/client/metrics.go:132]: Ahh thanks for catching this. As for the gauge @prestonvanloon is worried about the same. Gauge is needed to get a time series which you need if you want to collect the table above. Getting a histogram is only useful to get averages. This as gauge replicates the table in beaconcha.in. The prometheus dbase can then later be pruned can't it?

File: validator/client/metrics.go [hunk 2]
```diff
@@ -122,8 +177,8 @@ var (
 // and penalties over time, percentage gain/loss, and gives the end user a better idea
 // of how the validator performs with respect to the rest.
 func (v *validator) LogValidatorGainsAndLosses(ctx context.Context, slot uint64) error {
-	if slot%params.BeaconConfig().SlotsPerEpoch != 0 || slot <= params.BeaconConfig().SlotsPerEpoch {
-		// Do nothing unless we are at the start of the epoch, and not in the first epoch.
+	if (slot+1)%params.BeaconConfig().SlotsPerEpoch != 0 || slot <= params.BeaconConfig().SlotsPerEpoch {
+		// Do nothing unless we are at the end of the epoch, and not in the first epoch.
```
Comment (prestonvanloon) [inline validator/client/metrics.go:181]: Why change this?

File: validator/client/runner.go [hunk 3]
```diff
@@ -117,9 +117,6 @@ func run(ctx context.Context, v Validator) {
 			// Report this validator client's rewards and penalties throughout its lifecycle.
 			log := log.WithField("slot", slot)
 			log.WithField("deadline", deadline).Debug("Set deadline for proposals and attestations")
-			if err := v.LogValidatorGainsAndLosses(slotCtx, slot); err != nil {
-				log.WithError(err).Error("Could not report validator's rewards/penalties")
-			}
```
Comment (prestonvanloon) [inline validator/client/runner.go:122]: Why move this?

**Review by prestonvanloon — COMMENTED** [review]
There are a few changes that seem unrelated. Please explain additional changes

File: validator/client/runner.go (hunk 3 above)
Comment (potuz) [reply validator/client/runner.go:122]: This is part of #7708 so perhaps is better to keep the discussion there?

**Comment by potuz** [comment]
Fixes #7694

**Comment by rauljordan** [comment]
@terencechain PTAL

**Review by terencechain — APPROVED** [review]

**Pull request merged**